scarb/src/core/package/name.rs

```diff
@@ -88,9 +88,89 @@ impl PackageName {
             }
         }
 
+        if PackageName::is_keyword(&name) {
+            bail!(
+                "the name `{name}` cannot be used as a package name, \
+                names can't use cairo keywords see the full list at \
+                https://docs.cairo-lang.org/language_constructs/keywords.html and \
+                https://github.com/starkware-libs/cairo/blob/dc7f0d5ca76b7409816d99772033abf58975a872/crates/cairo-lang-parser/src/lexer.rs#L184-L209"
+            )
+        }
+
+        if ["test", "scarb"].contains(&name.as_ref()) {
+            bail!("the name `{name}` cannot be used as a package name")
+        }
+
         Ok(Self(name))
     }
 
+    /// Checks if name is a Cairo keyword
+    pub fn is_keyword(name: impl AsRef<str>) -> bool {
+        [
+            "as",
+            "assert",
+            "break",
+            "const",
+            "continue",
+            "do",
+            "dyn",
+            "else",
+            "enum",
+            "extern",
+            "false",
+            "fn",
+            "for",
+            "hint",
+            "if",
+            "impl",
+            "implicits",
+            "in",
+            "let",
+            "loop",
+            "macro",
+            "match",
+            "mod",
+            "move",
+            "mut",
+            "nopanic",
+            "of",
+            "pub",
+            "ref",
+            "return",
+            "self",
+            "static",
+            "static_assert",
+            "struct",
+            "super",
+            "trait",
+            "true",
+            "try",
+            "type",
+            "typeof",
+            "unsafe",
+            "use",
+            "where",
+            "while",
+            "with",
+            "yield",
+        ]
+        .contains(&name.as_ref())
+    }
+
+    /// Checks if name is restricted on Windows platforms
+    pub fn is_windows_restricted(name: impl AsRef<str>) -> bool {
+        [
+            "con", "prn", "aux", "nul", "com1", "com2", "com3", "com4", "com5", "com6", "com7",
+            "com8", "com9", "lpt1", "lpt2", "lpt3", "lpt4", "lpt5", "lpt6", "lpt7", "lpt8", "lpt9",
+        ]
+        .contains(&name.as_ref())
+    }
+
+    /// Checks if name equals `core` or `starknet`
+    pub fn is_internal(name: impl AsRef<str>) -> bool {
+        [CORELIB_CRATE_NAME, "starknet"].contains(&name.as_ref())
+    }
+
```

Comment:
nitpick: could we have these functions in separate module, as top-level funcs, like Cargo has? Would help reduce this file size.

I suggest `scarb::internal::restricted_names`